runtime/src/parachains.rs

```diff
@@ -144,49 +177,119 @@ decl_storage! {
+						);
+
+						// Either there are no more messages to add...
+						if !head.candidate.upward_messages.is_empty() {
+							let (count, size) = <RelayDispatchQueueSize<T>>::get(id);
+							ensure!(
+								// ...or we are appending one message onto an empty queue...
+								head.candidate.upward_messages.len() + count as usize == 1
+								// ...or...
+								|| (
+									// ...the total messages in the queue ends up being no greater than the
+									// limit...
+									head.candidate.upward_messages.len() + count as usize <= MAX_QUEUE_COUNT
+								&&
+									// ...and the total size of the payloads in the queue ends up being no
+									// greater than the limit.
+									head.candidate.upward_messages.iter()
+										.fold(size as usize, |a, x| a + x.1.len())
+									<= WATERMARK_QUEUE_SIZE
+								),
+								"Messages added when queue full"
+							);
+						}
+
+						Self::check_egress_queue_roots(&head, &active_parachains)?;
 
-			// perform integrity checks before writing to storage.
-			{
-				let n_parachains = active_parachains.len();
-				ensure!(heads.len() <= n_parachains, "Too many parachain candidates");
-
-				let mut last_id = None;
-				let mut iter = active_parachains.iter();
-				for head in &heads {
-					// proposed heads must be ascending order by parachain ID without duplicate.
-					ensure!(
-						last_id.as_ref().map_or(true, |x| x < &head.parachain_index()),
-						"Parachain candidates out of order by ID"
-					);
-
-					// must be unknown since active parachains are always sorted.
-					ensure!(
-						iter.find(|x| x == &&head.parachain_index()).is_some(),
-						"Submitted candidate for unregistered or out-of-order parachain {}"
-					);
-
-					Self::check_egress_queue_roots(&head, &active_parachains)?;
-
-					last_id = Some(head.parachain_index());
+						last_id = Some(head.parachain_index());
+					}
 				}
-			}
 
-			Self::check_attestations(&heads)?;
+				Self::check_attestations(&heads)?;
 
-			for head in heads {
-				let id = head.parachain_index();
-				<Heads<T>>::insert(id, head.candidate.head_data.0);
+				for head in heads.iter() {
+					let id = head.parachain_index();
+					<Heads<T>>::insert(id, &head.candidate.head_data.0);
+
+					// update egress.
+					for &(to, root) in &head.candidate.egress_queue_roots {
+						<Routing<T>>::insert((id, to), root);
+					}
+
+					// Queue up upwards messages (from parachains to relay chain).
+					if !head.candidate.upward_messages.is_empty() {
+						<RelayDispatchQueueSize<T>>::mutate(id, |&mut(ref mut count, ref mut len)| {
+							*count += head.candidate.upward_messages.len() as u32;
+							*len += head.candidate.upward_messages.iter()
+								.fold(0, |a, x| a + x.1.len()) as u32;
+						});
+						// Should never be able to fail assuming our state is uncorrupted, but best not
+						// to panic, even if it does.
+						let _ = <RelayDispatchQueue<T>>::append(id, &head.candidate.upward_messages[..]);
```

Comment:
A failure here should at least be logged, so that someone will notice it.

Reply:
logging makes little sense in the runtime. it would simply spam all full nodes.